File: src/javaasync/escalation/UnknownCollaborationEscalation.java
package javaasync.escalation;

import javaasync.Employee;

/**
 * Unknown collaboration escalation.
 */
public class UnknownCollaborationEscalation extends EscalationReport
{

    private final Employee sender;
    private final Employee receiver;

    /**
     * The constructor.
     *
     * @param sender - message sender.
     * @param receiver - message receiver.
     */
    public UnknownCollaborationEscalation(Employee sender, Employee receiver)
    {
        this.sender = sender;
        this.receiver = receiver;
    }

    /**
     * Handle unknown collaboration.
     */
    @Override
    public void handle()
    {
        if (sender.leavingJob() || receiver.leavingJob())
        {
            /*
             * Do nothing. It is not good to setup collaboration
             * if one of employees quit.
             */
        }
        else
        {
            sender.setupCollaboration(receiver);
        }

        recovered = true;
    }

    /**
     * Get reason in textual form.
     *
     * @return reason.
     */
    @Override
    public String reason()
    {
        return "Unkown collaboration between " + sender.getName() + " and "
            + receiver.getName()
            + ((recovered)
                ? " -> RECOVERED"
                : " -> NOT RECOVERED");
    }

}

File: src/javaasync/escalation/UnknownCollaborationEscalation.kt
package javaasync.escalation

import javaasync.AsyncService

/**
 * Unknown collaboration escalation.
 */
class UnknownCollaborationEscalation(
    private val sender: AsyncService,
    private val receiver: AsyncService
) : EscalationReport() {

    /**
     * Handle unknown collaboration.
     */
    override fun handle() {
        if (sender.leavingJob() || receiver.leavingJob()) {
            /*
             * Do nothing. It is not good to setup collaboration
             * if one of employees quit.
             */
        } else {
            sender.setupCollaboration(receiver)
        }

        recovered = true
    }

    /**
     * Get reason in textual form.
     *
     * @return reason.
     */
    override fun reason(): String {
        return ("Unkown collaboration between " + sender.name + " and "
                + receiver.name
                + if (recovered)
            " -> RECOVERED"
        else
            " -> NOT RECOVERED")
    }

}
